rust/index/src/hnsw_provider.rs

```diff
@@ -286,84 +305,92 @@ impl HnswIndexProvider {
-                                return Err(Box::new(HnswIndexProviderFileError::StorageError(e)));
-                            }
-                        };
-                        tracing::info!(
-                            "Fetched {} bytes from s3 for storage key {:?}",
-                            bytes_read,
-                            key,
-                        );
-                        Ok(buf)
-                    }).await
-                }
-            });
-        let buffers = try_join_all(futures).await?;
-        match hnswlib::HnswDataBuilder::new()
-            .header_buffer(buffers[0].clone())
-            .data_level0_buffer(buffers[1].clone())
-            .length_buffer(buffers[2].clone())
-            .link_list_buffer(buffers[3].clone())
-            .build()
-        {
-            Ok(hnsw_data) => Ok(hnsw_data),
-            Err(e) => Err(Box::new(HnswIndexProviderFileError::BuildError(e))),
-        }
+        hnsw_data_processor: FinishFn,
+    ) -> Result<FinishRet, chroma_storage::StorageError>
+    where
+        FinishFn: FnOnce(hnswlib::HnswData) -> FetchFut + Send + 'static,
+        FetchFut: Future<Output = Result<FinishRet, chroma_storage::StorageError>> + Send + 'static,
+        FinishRet: Clone + Sync + Send + 'static,
+    {
+        let keys: Vec<String> = FILES
+            .iter()
+            .map(|s| Self::format_key(prefix_path, source_id, s))
+            .collect();
+        let key_refs: Vec<&str> = keys.iter().map(|s| s.as_str()).collect();
+        let s3_fetch_span =
+            tracing::trace_span!(parent: Span::current(), "Read hnsw files from s3 into index");
+        let result = self
+            .storage
+            .fetch_batch(
+                key_refs,
+                GetOptions::new(StorageRequestPriority::P0).with_parallelism(),
+                |bufs| async move {
+                    assert_eq!(bufs.len(), 4);
+                    let unwrapped_buffers = bufs
+                        .iter()
+                        .map(|wrapped| {
+                            wrapped.as_ref().map_err(|e| {
+                                chroma_storage::StorageError::CallbackError {
+                                    info: e.to_string(),
+                                }
+                            })
+                        })
+                        .collect::<Result<Vec<_>, _>>()?;
+                    let hnsw_data = hnswlib::HnswData::builder()
+                        .header_buffer(unwrapped_buffers[0].clone())
+                        .data_level0_buffer(unwrapped_buffers[1].clone())
+                        .length_buffer(unwrapped_buffers[2].clone())
+                        .link_list_buffer(unwrapped_buffers[3].clone())
+                        .build();
+                    if let Ok(hnsw_data) = hnsw_data {
+                        hnsw_data_processor(hnsw_data).await
```

Comment:
why this closure as opposed to directly calling a method? Seems unnecessary, no?